src/material_library.h

```diff
@@ -0,0 +1,173 @@
+  */
+  void append_to_nuclist(pyne::Material mat);
+  /**
+   * \brief determines in that datapath exists in the hdf5 file
+   * \param[in] filename of the h5m file
+   * \param[in] the datapath of we would like to test
+   * \return true/false
+   */
+  bool hdf5_path_exists(const std::string& filename,
+                        const std::string& datapath);
+
+  /**
+   * \brief determines the length of an hdf5 data table
+   * \param[in] filename of the h5m file
+   * \param[in] the datapath of we would like to read
+   * \return the number of elements to the array
+   */
+  int get_length_of_table(const std::string& filename,
+                          const std::string& datapath);
+
+ public:
+  // materialLibrary constructor
+  MaterialLibrary();  //< empty constryctor
+
+  /**
+   * Constructor from file
+   * \param filename path to file on disk, this file may be either in plaintext
+   *                 or HDF5 format.
+   * \param datapath Path to the materials in the file.
+  */
+  MaterialLibrary(const std::string& filename,
+                  const std::string& datapath);
+
+  ~MaterialLibrary() {};  //< default destructor
+
+  /**
+   * \brief loads the pyne materials in map of name vs Material
+    /// \param filename Path on disk to the HDF5 file.
+    /// \param datapath Path to the materials in the file.
+    /// \param protocol Flag for layout of material on disk.
+  */
+  void from_hdf5(const std::string& filename,
+                 const std::string& datapath);
+  /**
+   * \brief loads the pyne materials in map of name vs Material
+    /// \param filename Path on disk to the json file.
+  */
+  void from_json(const std::string& filename);
+  void load_json(Json::Value json);
+  Json::Value dump_json();
+  void write_json(const std::string& filename);
+  /**
+   * Writes MaterialLibrary out to an HDF5 file.
+   *  This happens according to protocol 1.
+   *  Writting in a file already containing 
+   *  This might cause issue to 
+   *  read material already present in the datapath, 
+   *  and for the new materials if written in the 
+   *  same datapath.
+   *  \param filename Path on disk to the HDF5 file.
+   *  \param datapath Path to the the material in the file.
+  */
+  void write_hdf5(const std::string& filename,
+                  const std::string& datapath = "/materials");
+  /// Writes this nucpath to an HDF5 file.
+  /// This happens according to protocol 1.
+  /// \param db HDF5 id for the open HDF5 file.
+  /// \param nucpath Path to the nuclides list in the file.
+  void write_hdf5_nucpath(hid_t db, std::string nucpath);
+  /**
+   * \brief Merge a material library into the current one
+   * \param mat_library pyne material library to merge
+  */
+  void merge(pyne::MaterialLibrary mat_lib);
+  void merge(pyne::MaterialLibrary* mat_lib);
+
+  /**
+   * \brief Add a material to the library
+   * \param mat material to add
+  */
+  void add_material(pyne::Material mat);
```

Comment:
const reference?

Reply:
same as below:
I did not because we ensure that the material have a material_number.
otherwise I need to make a local copy within the method.